mara_pipelines/pipelines.py

```diff
@@ -81,24 +81,49 @@ def html_doc_items(self) -> List[Tuple[str, str]]:
 
 
 class Task(Node):
-    def __init__(self, id: str, description: str, commands: Optional[List[Command]] = None, max_retries: Optional[int] = None) -> None:
+    def __init__(self, id: str, description: str, commands: Optional[Union[Callable, List[Command]]] = None, max_retries: Optional[int] = None) -> None:
         super().__init__(id, description)
-        self.commands = []
+        self.callable_commands = callable(commands)
         self.max_retries = max_retries
 
-        for command in commands or []:
-            self.add_command(command)
-
-    def add_command(self, command: Command, prepend=False):
+        if self.callable_commands:
+            self._commands = None
+            self.__commands_callable = commands
+        else:
+            self._commands = []
+            self._add_commands(commands or [])
+
+    def _test_is_not_dynamic(self):
+        if self.callable_commands:
+            raise Exception('You cannot use add_command when the task is constructed with a callable commands function.')
+
+    @property
+    def commands(self) -> List:
+        if not self._commands:
```

Comment:
This will error in case you pass zero commands in: the if will assume no _commands and call the function which is not defined.

-> `if self._commands is None:`

Reply:
done